Backup: Add backup storage UI

Initsogar opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
- [x] Add site storage size and storage limit on backup admin page
- [x] Add progress bar with different colors based on % of storage usage
- [x] Adjust new layout to include the new storage UI
- [x] When storage usage is above 65%, show a CUT with add-on storage upgrade

#### What's next?
The following features will be address in a new PR:
- Add days of backups saved below the storage meter.
- Backups stopped related messages on main title and CUT.

#### Screenshots
| Description | Screenshot |
|---|---|
| Admin page | <img width="500" alt="image" src="https://user-images.githubusercontent.com/1488641/213733024-cb771c7c-3abc-4083-b77a-41ef57bef775.png">
| Normal (< 65% used storage) | <img src="https://user-images.githubusercontent.com/1488641/213374355-8c74fc03-7e68-4ccc-aeed-bc3448ba6426.png" width="500"> |
| Warning (65% - 80% used storage) | <img src="https://user-images.githubusercontent.com/1488641/213373897-ee64af55-6188-4890-a94c-31caa990526f.png" width="500"> | 
| Critical (80%-100% used storage) | <img src="https://user-images.githubusercontent.com/1488641/213732435-c61b3142-c7cb-4572-b950-a6cdb6ef8b7c.png" width="500"> |
| Full (> 100% used storage) | <img src="https://user-images.githubusercontent.com/1488641/213374019-bac94275-5a41-4838-aa28-6385bff5be6a.png" width="500"> |

#### Other information:
- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
* peaFOp-wy-p2
* p1HpG7-jQY-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

#### Unit testing:
You can run unit testing using the following command:
* JS: `jetpack test packages/backup js`
* PHP: `jetpack test packages/backup php`

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
* Spin up a Jurassic Ninja site with Jetpack Beta, using Backup branch `add/backup-storage-ui` with a backup plan like Jetpack Backup (1 GB or 10 GB).
* Navigate to Jetpack -> VautlPress Backup
* Once the first backup completes, you should see the main page of the plugin, including the storage meter.
* Given it is a new site, it will be ~0GB used and no upsell prompt should be displayed.

To try different scenarios for the storage meter, you could try to upload big files to your site or dispatch the events using [Redux DevTools (chrome extension)](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en). After installing it, follow these steps:
* Open the extension in Chrome
* In the dropdown in the top right, select `jetpack-backup-plugin`
* In the bottom, click on `>_` option to show the dispatcher.
![image](https://user-images.githubusercontent.com/1488641/213371014-0eaa92d6-d057-41ea-8d1f-2f38a26d026e.png)
* It should look like this:
<img src="https://user-images.githubusercontent.com/1488641/213371077-5db3dc19-35cb-4210-9075-71ea2ef236f0.png" width="250">

* Then in the dispatcher editor, you can set the used storage like this (the size is represented in bytes):
```javascript
{
    type: 'SITE_BACKUP_SIZE_GET_SUCCESS',
    payload: {
        size: 10737418240
    }
}
```

* So if you have a site with 10 GB, you can use the following sizes to test the different scenarios:
  * Normal (< 65%): 1073741824 (1GB)
  * Warning (65% - 80%): 7516192768 (7GB)
  * Critical (80%-100%): 9663676416 (9GB)
  * Full (> 100%): 10737418240 (10 GB)

Initsogar committed:
Install dependency for ProgressBar

Initsogar committed:
Add get site size and policies APIs

Initsogar committed:
Add actions, reducers and selectors for rewind details

Initsogar committed:
Add BackupStorageSection component for Storage UI

Initsogar committed:
Adjust admin layout to include BackupStorageSection

Initsogar committed:
Fix review request component width

Initsogar committed:
changelog

Initsogar committed:
Move backup storage section on top for lower resolutions

Initsogar committed:
Update backup package version to 1.11.0-alpha and composer.lock

Initsogar committed:
Merge branch 'trunk' into add/backup-storage-ui-test

Initsogar committed:
Merge branch 'trunk' into add/backup-storage-ui

Initsogar committed:
[not verified] Merge branch 'trunk' into add/backup-storage-ui

Initsogar committed:
UI improvements

Initsogar committed:
Storage addon plugin API

Initsogar committed:
Refactor add storage UI and add upsell prompt

keoshi reviewed (dismissed):
First off, thank you so much for all the work that went into this, @Initsogar! The detailed instructions are wonderful as well.

Everything looks great to me in the testing that I did. Noticed some minor things which I'll list below, but none are blocking; more like nice to haves if we want to create follow-up PRs.

#### Zero character shows up — bug?

While the first backup is being created a zero (`0`) appears in the section created for the storage UI; see below:

![image](https://user-images.githubusercontent.com/390760/213437933-5cbc82be-988f-44c0-ba2c-08b8787a478a.png)

#### Grid positioning

The new storage UI seems to take 6 columns of the existing grid, instead of the last 5 as intended. This one is on me for not having documented this more in detail...sorry about that.

See the following image where the top is the design and below what's in the PR right now, with a red rectangle of where the storage UI should be placed.

![image](https://user-images.githubusercontent.com/390760/213440708-a8e9485a-2299-4e5d-9d46-7d2d5a9a16ee.png)

This might seem confusing or counter-intuitive, but by adding a full empty column in between the left and right sections, we're creating more opportunities to draw the eye to the storage UI.

#### Heading in warning state

Saw that you added variable headings for different states — very nice touch. The warning state is missing a custom heading though. Wonder if we could add this one:

> Cloud storage is almost full

Initsogar committed:
Fix BackupStorageSpace returning 0 instead of not rendering

Initsogar committed:
Adjust BackupStorageSpace grid positioning

Initsogar commented:
@keoshi thank you for the valuable feedback! Here's my comments:

- [x] **Zero character shows up — bug?**
It was showing a 0 when the component wasn't rendering. Just fixed it.

- [x] **Fix grid positioning**
Thanks for clarifying, for some reason my eyes decided to omit that empty column 😅. Fixed!

- [ ] **Heading in warning state**
Currently, I have set to display `Cloud storage full` when the storage usage is 100% or over; otherwise it would be `Cloud storage space`.
I was thinking that `Cloud storage is almost full` fits better for storage usage between 80%-100% (Critical). What do you think? Maybe we can include it as well for 65%-80% (Warning) and Critical.

macbre commented on projects/packages/backup/src/class-jetpack-backup.php:
```diff
@@ -480,6 +524,142 @@ public static function manage_dismissed_backup_review_request( $request ) {
+	 * @return string Suggested addon storage slug
+	 */
+	public static function get_storage_addon_upsell_slug( $bytes_used, $bytes_available ) {
+		$bytes_10gb  = 10737418240;
+		$bytes_100gb = 107374182400;
+		$bytes_1tb   = 1099511627776;
```
Can we leave a comment here explaining where these IDs are coming from?

pypt replied:
I think those are not IDs but rather byte sizes. One way to remake this into a more readable form would be:

```php
		$bytes_10gb  = 10 * 1024 * 1024 * 1024;
		$bytes_100gb = 100 * 1024 * 1024 * 1024;
		$bytes_1tb   = 1024 * 1024 * 1024 * 1024;
```

macbre replied:
You're right 🤦🏻  But still, worth making these more readable.

Initsogar replied:
Thanks @macbre and @pypt for the recommendation.
Just made it a bit clearer, could you [take a look](https://github.com/Automattic/jetpack/pull/28085/files#diff-d8bc480adb260c41153b4d6a0614a3b5e8eb0b98cc6a151b19aa308c66ccaa05R586)?

macbre replied:
Way better now!

macbre commented on projects/packages/backup/src/class-jetpack-backup.php:
```diff
@@ -480,6 +524,142 @@ public static function manage_dismissed_backup_review_request( $request ) {
+	 * @param int $bytes_available Storage limit.
+	 * @return string Suggested addon storage slug
+	 */
+	public static function get_storage_addon_upsell_slug( $bytes_used, $bytes_available ) {
```
Some tests coverage would be nice here 🙂

Initsogar replied:
@macbre I just added some tests for this function.
Could you [take a look](https://github.com/Automattic/jetpack/pull/28085/files#diff-f0772d597476355e68faa1da0c2f88871f9856040fe8d3157711ec2021c353d9) and let me know what do you think?

macbre replied:
Perfect! 👌🏻

keoshi commented:
> I was thinking that Cloud storage is almost full fits better for storage usage between 80%-100% (Critical).

@Initsogar This sounds wonderful to me!

Initsogar committed:
Merge branch 'trunk' of github.com:Automattic/jetpack into add/backup-storage-ui

Initsogar committed:
Add a different section header when usage is critical

Initsogar commented:
> > I was thinking that Cloud storage is almost full fits better for storage usage between 80%-100% (Critical).
> 
> @Initsogar This sounds wonderful to me!

@keoshi Great! Thanks for the feedback.

- [x] **Heading in warning state**
Just added the `Cloud storage is almost full` when the storage usage is 80%-100% (Critical).

<img width="453" alt="image" src="https://user-images.githubusercontent.com/1488641/213732435-c61b3142-c7cb-4572-b950-a6cdb6ef8b7c.png">

Initsogar committed:
Make addon storage  sizes more readable

Initsogar committed:
Unit testing for storage addon upsell

Initsogar committed:
Access class const property with `self`

macbre commented on projects/packages/backup/tests/php/test-storage-addon-upsell.php:
```diff
@@ -0,0 +1,50 @@
+ *
+ * @package automattic/jetpack-backup
+ */
+class Test_Storage_Addon_Upsell extends TestCase {
```
Thanks for adding the test suite.

Initsogar committed:
Refactor, replacing `rewind` for `backup`

jwebbdev commented on projects/packages/backup/src/js/components/backup-storage-space/index.jsx:
```diff
@@ -0,0 +1,85 @@
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useState, useEffect, useMemo } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import React from 'react';
```
Do we need to import React here?

Initsogar replied:
@jwebbdev No, thanks for catching it! Just removed it.

jwebbdev commented on projects/packages/backup/src/js/components/backup-storage-space/storage-usage-details/use-storage-usage-text.tsx:
```diff
@@ -0,0 +1,63 @@
+import { useMemo, createInterpolateElement } from '@wordpress/element';
+import { __, sprintf } from '@wordpress/i18n';
+import React from 'react';
```
Do we need to import React here?

Initsogar replied:
Same as above, just removed it :)

Initsogar committed:
Remove unused import

Initsogar committed:
Merge branch 'trunk' of github.com:Automattic/jetpack into add/backup-storage-ui

Initsogar committed:
Merge branch 'trunk' of github.com:Automattic/jetpack into add/backup-storage-ui

jwebbdev reviewed (dismissed):
Looks good to me!

Initsogar committed:
Merge branch 'trunk' into add/backup-storage-ui

anomiex commented on projects/packages/backup/package.json:
```diff
@@ -26,7 +26,9 @@
 		"extends @wordpress/browserslist-config"
 	],
 	"dependencies": {
+		"@automattic/components": "2.0.1",
 		"@automattic/jetpack-analytics": "workspace:*",
+		"@automattic/format-currency": "1.0.1",
```
Alphabetical order is nice to maintain.
```suggestion
		"@automattic/components": "2.0.1",
		"@automattic/format-currency": "1.0.1",
		"@automattic/jetpack-analytics": "workspace:*",
```

Initsogar replied:
Thanks for the suggestion! Just commited it 😊

anomiex reviewed (commented):
One minor suggestion inline. Otherwise looks ok to me from a monorepo perspective.

Initsogar committed:
Sort dependencies in alphabetical order

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into add/backup-storage-ui

Initsogar committed:
Merge branch 'trunk' into add/backup-storage-ui
